### ISSUE TITLE
Retrieving a user by specified keys are being ignored when a JWT cookie is available. 

### ISSUE THREAD

**Issue by rfg-anthony-lim** [FusionAuth/fusionauth-issues#43]
### Description 
I installed version 1.2.1
https://fusionauth.io/docs/v1/tech/apis/users#retrieve-a-user
When trying to retrieve User endpoints such as below, they only return the user profile of the JWT cookie user; they do not return the user profile of the requested user key:
/api/user/{userId}
/api/user?loginId={loginId}
/api/user?email={email}
/api/user?username={username}

### Steps to reproduce
1. Authenticate using /api/login using your API client. I use Insomnia app. 
```
> POST /api/login HTTP/1.1
> Host: xxx.xxx.xxx.xxx:9011
> User-Agent: insomnia/6.2.0
> Content-Type: application/json
> X-Forwarded-For: ::
> X-FusionAuth-TenantId: <My Tenant Uuid>
> Accept: */*
> Content-Length: 268
| {
| 	"loginId": "user1",
| 	"password": "password",
| 	"ipAddress": "127.0.0.1",
| 	"device": "XXX",
| 	"metaData": {
| 		"device": {
| 			"description": "Insomnia App",
| 			"lastAccessedAddress": "127.0.0.1",
| 			"name": "Insomnia",
| 			"type": "OTHER"
| 		}
| 	}
| }
< HTTP/1.1 212 access_token="eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1NDI4NTk4MzIsImlhdCI6MTU0Mjg1NjIzMiwiaXNzIjoiMTcyLjE2LjAuMjIyOjkwMTEiLCJzdWIiOiJlNWZiNmE5Ny01ZWFmLTRjZjMtODJhYy1mNDMzOWUyM2NkNzAiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoidXNlcjFAcmVkZmxhZ2dyb3VwLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZX0._kUQmTKFdG1PoO1BOm9F6Sk_L7k34R83hQSDodMHqbU" for domain xxx.xxx.xxx.xxx, path /, expire 0
< Set-Cookie: access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1NDI4NzQ2NDQsImlhdCI6MTU0Mjg3MTA0NCwiaXNzIjoiMTcyLjE2LjAuMjIyOjkwMTEiLCJzdWIiOiI1NDcyYzU3MC0zOTJiLTQxOTYtYTdmNC0xYzg4Yjc0NDEwNzEiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiZnVzaW9uYXV0aEBtYWlsbmVzaWEuY29tIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlfQ.SaVgDC9xCM154crNCnOSGVG7qoR5BOkpblaZz1tSeo8; Path=/; HttpOnly
< Content-Type: application/json;charset=UTF-8
< Content-Length: 907
< Date: Thu, 22 Nov 2018 07:17:24 GMT

* Received 833 B chunk
* Connection #3 to host xxx.xxx.xxx.xxx left intact
* Saved 1 cookie
```
2. Retrieve user by Id /api/user/25a872da-bb44-4af8-a43d-e7bcb5351ebc; It doesn't have to be an existing User UUID. Your API Client should be sending the token cookie automatically. 
```
> GET /api/user/25a872da-bb44-4af8-a43d-e7bcb5351ebc HTTP/1.1
> Host: xxx.xxx.xxx.xxx:9011
> User-Agent: insomnia/6.2.0
> Cookie: access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE1NDI4NzQ2NDQsImlhdCI6MTU0Mjg3MTA0NCwiaXNzIjoiMTcyLjE2LjAuMjIyOjkwMTEiLCJzdWIiOiI1NDcyYzU3MC0zOTJiLTQxOTYtYTdmNC0xYzg4Yjc0NDEwNzEiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiZnVzaW9uYXV0aEBtYWlsbmVzaWEuY29tIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlfQ.SaVgDC9xCM154crNCnOSGVG7qoR5BOkpblaZz1tSeo8
> X-Forwarded-For: ::
> X-FusionAuth-TenantId: <My Tenant Uid>
> Authorization: <My API Key>
> Accept: */*
< HTTP/1.1 200 
< Content-Type: application/json;charset=UTF-8
< Content-Length: 553
< Date: Thu, 22 Nov 2018 07:21:18 GMT

* Received 553 B chunk
* Connection #36 to host 172.16.0.222 left intact
{
  "user": {
    "active": true,
    "birthDate": "1914-11-21",
    "email": "fusionauth@mailnesia.com",
    "expiry": 1543507200000,
    "firstName": "Mail",
    "fullName": "Mail Nesia",
    "id": "5472c570-392b-4196-a7f4-1c88b7441071",
    "insertInstant": 1542768797248,
    "lastLoginInstant": 1542871044096,
    "lastName": "Nesia",
    "mobilePhone": "+85254000455",
    "passwordChangeRequired": false,
    "passwordLastUpdateInstant": 1542769401224,
    "tenantId": "e360a05d-ef17-4336-a82f-b3e69320d036",
    "twoFactorDelivery": "TextMessage",
    "twoFactorEnabled": true,
    "username": "fusionauth",
    "usernameStatus": "ACTIVE",
    "verified": false
  }
}
```
3. Look at the response body. It returns the user profile of the JWT owner, not 404 or the profile of the specified UUID. 

### Expected behavior
The API should return the user profile of the specified key -- uuid, loginId, email, username, etc.

### Screenshots
No screenshots. I've already sent the detailed HTTP communication. 

### Platform
 - Device: Desktop and Laptop
 - OS: OS: Windows 10 (Client), Ubuntu 18.10 (Server)
 - Insomnia 6.2.0

### Additional context
It seems that the path or request arguments are being ignored when a JWT cookie exists.

**Comment by robotdan** [comment]
Thanks @rfg-anthony-lim for the excellent details. 

This is the currently intended behavior for this API. The intention of this design was that if you are logged into FusionAuth you may retrieve your own user profile. In this scenario, regardless of the UUID you provide on the URL it will be ignored and replaced with the `sub` claim from the provided JWT. 

I can review this design with team and see how to handle this issue.

A couple of options exist to address this issue:

1. Prefer the API key if present, and in that case ignore the JWT (provided as a cookie or an Auth header)
2. If a UUID is requested other than that matching the `sub` claim in the JWT, return a `404` or perhaps more appropriate a `401` or `403`. 
3. Remove the option to use a JWT with this API as a form of authentication
4. No change

I can see why this behavior can be confusing. However, is this a real use case you plan to utilize? 

For example, will you have a user :
1. Log into FusionAuth with a user-agent based application which will store a cookie
2. Make an API call to the User API from that same user-agent based application using an API key 

Generally speaking because a user-agent based application is not considered secure, API keys are not stored or utilized in this domain.

In my opinion this scenario is not likely, but I may be missing another use case, so if you have another one in mind please describe it for me so I can better prioritize this issue.

**Comment by rfg-anthony-lim** [comment]
Thanks for the detailed reply, @robotdan .

I understand how my expected behaviour may open some security holes. So based on your options, we can implement the following*:
1. Combine 1 and 3
2. As usual, return `401` if API key is not present; `403` if access to these endpoints are disabled; `404` if not existing, etc. 

* Only applied to User endpoints that accepts specified keys:
/api/user/{userId}
/api/user?loginId={loginId}
/api/user?email={email}
/api/user?username={username}
/api/user?changePasswordId={changePasswordId}
/api/user?verificationId={verificationId}

The example you provided is precisely what I am working on. 

I agree that user-agent based applications are not considered secure (unless communicated on TLS/SSL, I may be wrong), and JavaScript encryption is hopeless. However, SPAs and front-end technologies are inevitable nowadays, and backend have become decoupled instances/services. We may consider the option to develop our own backend layer on top of FusionAuth to abstract details and keys, but not right now.

**Comment by robotdan** [comment]
Thanks for the additional information @rfg-anthony-lim. 

> I agree that user-agent based applications are not considered secure (unless communicated on TLS/SSL, I may be wrong), and JavaScript encryption is hopeless.

Hopefully TLS is a given, the problem you will have is that there is no place to securely store that API key in the browser. Perhaps your application is not going to be public and only accessible by trusted users? 

In any case, I would caution you with this approach, but if it fits your needs and you are ok with the risks, then feel free. 

Then change I plan to make to this API will be to prefer the API key over the JWT so that if you have a valid API key on the request you will be able to make the API call as you wish. 

Hopefully we can get this out in a patch release for you shortly.

**Comment by robotdan** [comment]
Internal: Need to verify the doc for each API that accepts a JWT to ensure that the order of precedence is correct. API key will not take precedence.

**Comment by robotdan** [comment]
Fixed in version 1.3.1.
